File: dogkin_ws/src/learning_service/scripts/turtle_spawn.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
# This program creates Client called "/turtle_spawn" for Service named "/spawn", datatype <turtlesim::Spawn>

import sys
import rospy
from turtlesim.srv import Spawn

def turtle_spawning():
	# create node	
	rospy.init_node('turtle_spawn')
	
	# wait for service /spawn of datatype <Spawn>	
	rospy.wait_for_service('/spawn')
	try:
		add_turtle = rospy.ServiceProxy('/spawn', Spawn)
		
		# when server is found, initialise service data (x, y, name)
		response = add_turtle(2.0, 2.0, 0.0, "turtle2")
		return response.name
	except rospy.ServiceException, e:
		print("Service call failed: %s" %e)


if __name__ == '__main__':
	# request for service and show result	
	print('Spawn turtle successfully [name:%s]' %(turtle_spawning()))

File: dogkin_ws/src/learning_service/src/turtle_spawn.cpp
/** 
* This program requests service named /spawn, service type turtlesim::Spawn
*/

#include <ros/ros.h>
#include <turtlesim/Spawn.h>

int main(int argc, char** argv)
{
	// initialise ROS node
	ros::init(argc, argv, "turtle_spawn");
	
	// create node handle
	ros::NodeHandle node;

	// search for service /spawn, create client [ros::ServiceClient] and connect to /spawn
	ros::service::waitForService("/spawn");
	ros::ServiceClient add_turtle = node.serviceClient<turtlesim::Spawn>("/spawn");

	// initialise turtlesim::Spawn service data
	turtlesim::Spawn srv;
	srv.request.x = 2.0;
	srv.request.y = 2.0;
	srv.request.name = "turtle2";

	// request for service
	ROS_INFO("Call service to spawn turtle[x:%0.6f, y:%0.6f, name:%s]", srv.request.x, srv.request.y, srv.request.name.c_str());

	add_turtle.call(srv);

	// show service call result
	ROS_INFO("Spawn turtle successfully [name:%s]", srv.request.name.c_str());

	return 0;
}
